### PR TITLE
Pr gpvi cleanup

### PR DESCRIPTION
**Pull request by runjerry**
Fixed a few bugs in GPVI and some cleanup

* Fixed output_dim in initializing the inverse_mvp in the generator for GPVI
* Fixed inverse_mvp_algorithm_test to ensure injectivity of the generator function
* Fixed the input dim mismatch for inverse_mvp in generator._rkhs_func_grad
* Removed unused files

### CODE REVIEW

**Review by emailweixu — DISMISSED** [review]

File: alf/algorithms/generator.py [hunk 1]
```diff
@@ -433,6 +423,55 @@ def __init__(self,
             self._predict_net_updater = common.get_target_updater(
                 self._net, self._predict_net, tau=net_moving_average_rate)
 
+    def _create_mvp_network(self, input_spec, output_dim, hidden_size,
+                            num_hidden_layers):
+        r"""Instantiate an encoding network to predict a matrix-vector
+        product of the form :math:`y=J^{-1}*vec, where :math:`J^{-1}` is a
+        square matrix and :math:`vec` is a vector. 
+        To train this mvp_network, a minimization objective is used
```
Comment (neale) [inline alf/algorithms/generator.py]: Right now these details overlap with the inverse_mvp_train_step function. Maybe we should keep one of the docstrings simple, right now both the network and the train_step function describe the minimization objective.

File: alf/algorithms/generator.py [hunk 2]
```diff
@@ -964,26 +967,23 @@ def _inverse_mvp_train_step(self, z, eps=None):
         generator :math:`f` and output :math`y`.
     
         Args: 
-            z (torch.tensor): of size [N2, K], representing z'
-            eps (torch.tensor): of size [N2, N, D or K], representing
-                :math:`\nabla_{z'}k(z', z)`, last dimension is D when 
-                self._force_fullrank is True, K otherwise. In general,
-                K is much less than D.
+            z (torch.tensor): of size [N2, D], representing z'
```
Comment (neale) [inline alf/algorithms/generator.py]: I think we want :math:\`z'\` to be consistent.

File: alf/algorithms/generator.py (hunk 1 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: Updated. moving the create_mvp_network into the init of InverseMVPAlgorithm and simplify docstring there.

File: alf/algorithms/generator.py (hunk 2 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: updated

File: alf/algorithms/generator.py [hunk 3]
```diff
@@ -119,43 +119,100 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
-            inputs (tuple of Tensors): inputs (z, eps) for prediction.
+            inputs (tuple of Tensors): inputs (z, vec) for prediction.
+            - z (Tensor): of size [N2, K] or [N2, D], representing :math:`z'`
+            - vec (Tensor): of size [N2, D] or [N2, N, D], representing 
+                :math:`\nabla_{z'}k(z', z)`.
             state: not used.
             
         Returns:
             AlgStep:
             - output (torch.Tensor): predictions of InverseMVP network
```
Comment (emailweixu) [inline alf/algorithms/generator.py]: Now the output also includes z_inputs. But why did you make this change?

File: alf/algorithms/generator.py [hunk 4]
```diff
@@ -119,43 +119,100 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
+        if vec.ndim == 2:
+            vec_inputs = vec
+        elif vec.ndim == 3:  # [N2, N, D]
+            z_inputs = torch.repeat_interleave(
+                z_inputs, vec.shape[1], dim=0)  # [N2*N, K]
+            vec_inputs = vec.reshape(vec.shape[0] * vec.shape[1],
+                                     -1)  # [N2*N, D]
+        else:
+            raise ValueError(
+                "vec must of dimension 3, got vec of dimension {}".format(
```
Comment (emailweixu) [inline alf/algorithms/generator.py]: vec must of => vec must be

File: alf/algorithms/generator.py [hunk 5]
```diff
@@ -119,43 +119,100 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
+            last_layer_size=output_dim,
+            last_activation=math_ops.identity,
+            name='InverseMVPNetwork')
 
     def predict_step(self, inputs, state=None):
         """Predict for one step of inputs.
         Args:
-            inputs (tuple of Tensors): inputs (z, eps) for prediction.
+            inputs (tuple of Tensors): inputs (z, vec) for prediction.
+            - z (Tensor): of size [N2, K] or [N2, D], representing :math:`z'`
```
Comment (emailweixu) [inline alf/algorithms/generator.py]: what is K? is it self._z_dim? please comment.

File: alf/algorithms/generator.py [hunk 6]
```diff
@@ -119,43 +119,100 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
             - state: not used.
         """
-        outputs = self._net(inputs)[0]
+        z_inputs, vec = inputs
+        assert z_inputs.ndim == 2 and z_inputs.shape[-1] >= self._z_dim
+        assert vec.shape[-1] == self._vec_dim
+        assert z_inputs.shape[0] == vec.shape[0]
+
+        if z_inputs.shape[-1] > self._z_dim:
+            z_inputs = z_inputs[:, self._z_dim]  # [N2, K]
```
Comment (emailweixu) [inline alf/algorithms/generator.py]:
```python
z_inputs = z_inputs[:, :self._z_dim]
```
I wonder how the unit test can still pass with this bug.

File: alf/algorithms/generator.py (hunk 3 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: Updated the output. For the z_inputs, it is a bit tedious to explain here, lets discuss it during lunch meeting.

File: alf/algorithms/generator.py (hunk 4 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: Updated.

File: alf/algorithms/generator.py (hunk 5 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: Yes. Updated.

File: alf/algorithms/generator.py (hunk 6 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: Oops...updated.
It ends up no tests ever getting into this case. 
- inverse_mvp_algorithm_test has K=2, D=3, but sends in z_inputs of shape [5, 2] since the Jacobian does not depend on the last component. Just revised it to [5, 3] to test this case.
- generator tests only test for K=D=2, I will explain to you the reason we did not include K<D for generator tests.

File: alf/algorithms/generator.py [hunk 7]
```diff
@@ -119,47 +119,102 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
-    If using ``svgd3``, then the eps quantity represents the kernel grad
-        :math:`\nabla_{z'}k(z', z)`
+
+    Maintain an encoding network that takes (z, vec) as input and predicts a 
+    matrix-vector product (mvp) of the form :math:`y=J^{-1}(z)*vec`, where 
+    :math:`J^{-1}(z)` is the inverse of the Jacobian matrix of some function
+    :math:`f(z)`, and ``vec`` is a vector. This network is used in GPVI in 
+    computing the ``functional_gradient`` of the generator, where :math:`J^{-1}`
+    is the inverse of the Jacobian of the generator function w.r.t. input noise
+    :math:`z'`, and ``vec`` is the gradient of the kernel :math:`\nabla_{z'}(z', z)`.
```
Comment (neale) [inline alf/algorithms/generator.py]: \nabla_{z'}(z', z) -> \nabla_{z'} k(z', z')

File: alf/algorithms/generator.py [hunk 8]
```diff
@@ -119,47 +119,102 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
+        if vec.ndim == 2:
+            vec_inputs = vec
+        elif vec.ndim == 3:  # [N2, N, D]
+            z_inputs = torch.repeat_interleave(
+                z_inputs, vec.shape[1], dim=0)  # [N2*N, K]
+            vec_inputs = vec.reshape(vec.shape[0] * vec.shape[1],
+                                     -1)  # [N2*N, D]
+        else:
+            raise ValueError(
+                "vec must be dimension 3, got vec of dimension {}".format(
```
Comment (neale) [inline alf/algorithms/generator.py]: vec can also be dimension 2 I believe.

File: alf/algorithms/generator.py [hunk 9]
```diff
@@ -119,47 +119,102 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
+        z_spec = TensorSpec(shape=(self._z_dim, ))
+        vec_spec = TensorSpec(shape=(self._vec_dim, ))
+
+        self._net = EncodingNetwork(
+            (z_spec, vec_spec),
+            input_preprocessors=(torch.nn.Linear(self._z_dim, hidden_size),
+                                 torch.nn.Linear(self._vec_dim, hidden_size)),
+            preprocessing_combiner=alf.layers.NestConcat(),
+            fc_layer_params=(2 * hidden_size, ) * num_hidden_layers,
+            activation=torch.relu_,
```
Comment (neale) [inline alf/algorithms/generator.py]: Maybe we should allow `activation` as an input argument. Not sure about this though.

**Review by neale — COMMENTED** [review]
Looks good to me, just some comments on docstrings.

File: alf/algorithms/generator.py (hunk 8 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: oh yeah...updated

File: alf/algorithms/generator.py (hunk 9 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: Agreed, updated.

File: alf/algorithms/generator.py (hunk 7 above)
Comment (runjerry) [reply alf/algorithms/generator.py]: updated

File: alf/algorithms/generator.py [hunk 10]
```diff
@@ -119,47 +119,105 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
         r"""Create a InverseMVPAlgorithm.
         Args:
-            output_dim (int): total output dimension of the inverse_mvp net, will differ
-                between ``svgd3`` and ``minmax`` methods
-            net (Network): network for predicting outputs from inputs.
+            input_dim (int): dimension of input z
+            output_dim (int): output dimension, i.e., dimension of the mvp 
+            hidden_size (int): width of hidden layers
+            num_hidden_layers (int): number of hidden layers after 
+            activation (nn.functional): activation used for all hidden layers.
```
Comment (emailweixu) [inline alf/algorithms/generator.py]: nn.functional => `Callable`. nn.functional is not a type.

Comment (runjerry) [reply alf/algorithms/generator.py]: Updated.

**Review by neale — APPROVED** [review]
Looks good to me
